In [1]:
import numpy as np
import pandas as pd

In [3]:
pd.__version__

'2.2.2'

In [22]:
# Import data into a df
df = pd.read_csv("https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv")

The dataframe has 9704 rows and 11 columns.

In [7]:
df.shape

(9704, 11)

In [9]:
df.head()

,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,model_year,origin,fuel_type,drivetrain,num_doors,fuel_efficiency_mpg
0,170,3.0,159.0,3413.433759,17.7,2003,Europe,Gasoline,All-wheel drive,0.0,13.231729
1,130,5.0,97.0,3149.664934,17.8,2007,USA,Gasoline,Front-wheel drive,0.0,13.688217
2,170,NaN,78.0,3079.038997,15.1,2018,Europe,Gasoline,Front-wheel drive,0.0,14.246341
3,220,4.0,NaN,2542.392402,20.2,2009,USA,Diesel,All-wheel drive,2.0,16.912736
4,210,1.0,140.0,3460.870990,14.4,2009,Europe,Gasoline,All-wheel drive,2.0,12.488369


In [11]:
df.fuel_type.value_counts()

fuel_type
Gasoline    4898
Diesel      4806
Name: count, dtype: int64

There are 2 type of fuels: Gasoline and diesel.

In [15]:
df.isna().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

4 columns have missing values: num_cylinders, horsepower, acceleration, and num_doors.

The maximum fuel efficiency (mpg) of cars with Asia origin is 23.75 mpg.

In [32]:
max(df[df['origin'] == 'Asia']['fuel_efficiency_mpg'])

23.759122836520497

In [36]:
df['horsepower'].median()

149.0

In [88]:
mode_horsepower = df.horsepower.mode()
mode_horsepower

0    152.0
Name: horsepower, dtype: float64

In [84]:
df['horsepower'] = df['horsepower'].fillna(df['horsepower'].mode()[0])

In [86]:
# value of horse power median after filling the missing values with most frequent value
df['horsepower'].median()

152.0

The value of median increased from 149 to 152.

## Question 7

In [101]:
asia_car_df = df[df['origin'] == 'Asia'][['vehicle_weight', 'model_year']].copy()

In [103]:
X = asia_car_df.head(7).values
X

array([[2714.21930965, 2016.        ],
       [2783.86897424, 2010.        ],
       [3582.68736772, 2007.        ],
       [2231.8081416 , 2011.        ],
       [2659.43145076, 2016.        ],
       [2844.22753389, 2014.        ],
       [3761.99403819, 2019.        ]])

In [105]:
# compute dot product of XTX
XTX = X.T.dot(X)
XTX

array([[62248334.33150762, 41431216.50732678],
       [41431216.50732678, 28373339.        ]])

In [107]:
# compute the inverse matrix of XTX
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 5.71497081e-07, -8.34509443e-07],
       [-8.34509443e-07,  1.25380877e-06]])

In [109]:
y = np.array([1100, 1300, 800, 900, 1000, 1100,1200])

In [111]:
w = (XTX_inv @ X.T) @ y

In [113]:
w

array([0.01386421, 0.5049067 ])

In [115]:
sum(w)

0.5187709081074007